docs/source/transform.rst

```diff
@@ -35,131 +35,201 @@ transformation, only one or all of the relevant methods
 
-.. autosummary::
+    class InsertCommentTransformation(Transformation):
 
-   loki.transform.transformation.Transformation
-   loki.transform.build_system_transform.CMakePlanner
-   loki.transform.build_system_transform.FileWriteTransformation
-   loki.transform.dependency_transform.DependencyTransformation
-   loki.transform.dependency_transform.ModuleWrapTransformation
-   loki.transform.fortran_c_transform.FortranCTransformation
-   loki.transform.fortran_max_transform.FortranMaxTransformation
-   loki.transform.fortran_python_transform.FortranPythonTransformation
-   loki.transform.transform_hoist_variables.HoistVariablesAnalysis
-   loki.transform.transform_hoist_variables.HoistVariablesTransformation
-   loki.transform.transform_hoist_variables.HoistTemporaryArraysAnalysis
-   loki.transform.transform_hoist_variables.HoistTemporaryArraysTransformationAllocatable
-   loki.transform.transform_parametrise.ParametriseTransformation
-
-Further transformations are defined for specific use-cases but may prove
-useful in a wider context. These are defined in :mod:`transformations`:
+        # When called on a Sourcefile, automatically apply this to all modules
+        # in the file
+        recurse_to_modules = True
 
-.. autosummary::
+        # When called on a Sourcefile or Module, automatically apply this to all
+        # Subroutines in the file or module
+        recurse_to_procedures = True
+
+        def _insert_comment(self, program_unit):
+            program_unit.spec.prepend(Comment(text='! Processed by Loki'))
+
+        def transform_subroutine(self, routine, **kwargs):
+            self._insert_comment(routine)
+
+        def transform_module(self, module, **kwargs):
+            self._insert_comment(module)
+
+With these two attributes added, we can now apply the transformation to all
+modules and procedures in a single command:
+
+.. code-block:: python
 
-   transformations.argument_shape.ArgumentArrayShapeAnalysis
-   transformations.argument_shape.ExplicitArgumentArrayShapeTransformation
-   transformations.data_offload.DataOffloadTransformation
-   transformations.derived_types.DerivedTypeArgumentsTransformation
-   transformations.dr_hook.DrHookTransformation
-   transformations.single_column_claw.ExtractSCATransformation
-   transformations.single_column_claw.CLAWTransformation
-   transformations.single_column_coalesced.SingleColumnCoalescedTransformation
-   transformations.scc_cuf.SccCufTransformation
+   source = Sourcefile(...)  # may contain modules and subroutines
+   transformation = InsertCommentTransformation()
+   transformation.apply(source)
 
-Additionally, a number of tools for common transformation tasks are provided as
-functions that can be readily used in a step of the transformation pipeline:
+Most transformations, however, will only require modifying those parts of a file
+that are part of the call tree that is to be transformed to avoid unexpected
+side-effects. For that reason,
```

Comment:
sentence not quite finished?

Reply:
Indeed, but don't remember where I was going with this...